### ISSUE TITLE
Mypy throws errors when MagicMock-ing a method

### ISSUE THREAD

**Issue by MatanRubin** [python/mypy#6713]
Using Python 3.7 and MyPy 0.700.

When MagicMocking a method and then using MagicMock methods, mypy complains that the mocked method does not have the attribute "assert_called" and other MagicMock methods.
I'm not sure if it's at all possible for mypy to understand a method was monkey-patched like this, but it makes checking tests involving Mocks with mypy is not practical. Ideally, this code should not throw any errors.

For example:
```
from unittest.mock import MagicMock


class Foo:
    def bar(self) -> None:
        pass


foo = Foo()
foo.bar = MagicMock()
foo.bar()
foo.bar.assert_called()
```

When running Mypy, you get:

```
test.py:10: error: Cannot assign to a method
test.py:12: error: "Callable[[], None]" has no attribute "assert_called"
```

The first error is already discussed in https://github.com/python/mypy/issues/2427.
I couldn't find an issue discussing the second error.

**Comment by ilevkivskyi** [comment]
I could see how this can be problematic, but this is hard to support. First, we would need to add at least some basic generic support for monkey-patching.

**Comment by graingert** [comment]
The context manager patch interface should work here:

```python
from unittest.mock import MagicMock


class Foo:
    def bar(self) -> None:
        pass


foo = Foo()
with mock.patch.object(foo, "bar") as bar:
    foo.bar()
bar.assert_called()
```

**Comment by david-shiko** [comment]
> The context manager patch interface should work here:
> 
> ```python
> from unittest.mock import MagicMock
> 
> 
> class Foo:
>     def bar(self) -> None:
>         pass
> 
> 
> foo = Foo()
> with mock.patch.object(foo, "bar") as bar:
>     foo.bar()
> bar.assert_called()
> ```

What if the mocked object is pytest fixture, how to specify the return type?

**Comment by vors** [comment]
A more complex cases are still not well-covered. For example when the mock setup is split into separate methods and so on.

**Comment by hauntsaninja** [comment]
With modern mypy and typeshed, you only get the "Cannot assign to a method". 

If the context manager solution graingert mentions doesn't work for you, and you'd like to avoid this error, use `--disable-error-code=method-assign` (this can be enabled on a per-file basis). See also https://github.com/python/mypy/issues/2427#issuecomment-1419206807
